src/builder/Actions.sol

```diff
@@ -819,6 +854,177 @@ library Actions {
+            paymentMaxCost: payment.isToken ? PaymentInfo.findMaxCost(payment, borrowInput.chainId) : 0
+        });
+
+        return (quarkOperation, action);
+    }
+
+    function morphoRepay(MorphoRepay memory repayInput, PaymentInfo.Payment memory payment)
+        internal
+        pure
+        returns (IQuarkWallet.QuarkOperation memory, Action memory)
+    {
+        bytes[] memory scriptSources = new bytes[](1);
+        scriptSources[0] = type(MorphoBlueActions).creationCode;
+
+        Accounts.ChainAccounts memory accounts =
+            Accounts.findChainAccounts(repayInput.chainId, repayInput.chainAccountsList);
+
+        Accounts.QuarkState memory accountState = Accounts.findQuarkState(repayInput.repayer, accounts.quarkStates);
+
+        Accounts.AssetPositions memory repayAssetPositions =
+            Accounts.findAssetPositions(repayInput.assetSymbol, accounts.assetPositionsList);
+
+        Accounts.AssetPositions memory assetPositions =
+            Accounts.findAssetPositions(repayInput.collateralAssetSymbol, accounts.assetPositionsList);
+
+        uint256 repayAmount = repayInput.amount;
+        uint256 repayShares = 0;
+
+        if (repayInput.payInShares) {
+            repayAmount = 0;
+            repayShares = repayInput.borrowedShares;
+        }
+        bytes memory scriptCalldata = abi.encodeWithSelector(
+            MorphoBlueActions.repayAndWithdrawCollateral.selector,
+            MorphoInfo.getMorphoAddress(),
+            MorphoInfo.getMarketParams(repayInput.chainId, repayInput.collateralAssetSymbol, repayInput.assetSymbol),
+            repayAmount,
+            repayShares,
+            repayInput.collateralAmount,
+            repayInput.repayer,
+            repayInput.repayer
+        );
+
+        // Construct QuarkOperation
+        IQuarkWallet.QuarkOperation memory quarkOperation = IQuarkWallet.QuarkOperation({
+            nonce: accountState.quarkNextNonce,
+            scriptAddress: CodeJarHelper.getCodeAddress(type(MorphoBlueActions).creationCode),
+            scriptCalldata: scriptCalldata,
+            scriptSources: scriptSources,
+            expiry: repayInput.blockTimestamp + STANDARD_EXPIRY_BUFFER
+        });
+
+        // Construct Action
+
+        // Single element arrays to fit to existing context format
+        uint256[] memory collateralAmountsArray = new uint256[](1);
+        collateralAmountsArray[0] = repayInput.collateralAmount;
+        string[] memory collateralAssetSymbolsArray = new string[](1);
+        collateralAssetSymbolsArray[0] = repayInput.collateralAssetSymbol;
+        uint256[] memory collateralTokenPricesArray = new uint256[](1);
+        collateralTokenPricesArray[0] = assetPositions.usdPrice;
+        address[] memory collateralTokensArray = new address[](1);
+        collateralTokensArray[0] = assetPositions.asset;
+
+        RepayActionContext memory repayActionContext = RepayActionContext({
+            amount: repayInput.amount,
```

Comment:
So for repay max, this will always be `type(uint256).max` right?

Reply:
Yes same a cometRepay, builder will sees it and apply full shares amount retrieved from account list.

Reply:
Now MorphoScripts.sol will take care of finding shares, instead of relying on accounts info.